x-pack/plugins/reporting/server/core.ts

```diff
@@ -43,33 +44,39 @@ interface ReportingInternalStart {
 export class ReportingCore {
   private pluginSetupDeps?: ReportingInternalSetup;
   private pluginStartDeps?: ReportingInternalStart;
-  private browserDriverFactory?: HeadlessChromiumDriverFactory;
-  private readonly pluginSetup$ = new Rx.ReplaySubject<ReportingInternalSetup>();
-  private readonly pluginStart$ = new Rx.ReplaySubject<ReportingInternalStart>();
+  private readonly pluginSetup$ = new Rx.ReplaySubject<boolean>(); // for pluginHasSetup
+  private readonly pluginStart$ = new Rx.ReplaySubject<ReportingInternalStart>(); // for getPluginStartDeps
   private exportTypesRegistry = getExportTypesRegistry();
+  private config?: ReportingConfig;
 
-  constructor(private config: ReportingConfig) {}
+  constructor() {}
 
   public pluginSetup(reportingSetupDeps: ReportingInternalSetup) {
     this.pluginSetupDeps = reportingSetupDeps;
-    this.pluginSetup$.next(reportingSetupDeps);
+    this.pluginSetup$.next(true);
   }
 
   public pluginStart(reportingStartDeps: ReportingInternalStart) {
     this.pluginStart$.next(reportingStartDeps);
   }
 
-  public pluginHasStarted(): Promise<boolean> {
-    return this.pluginStart$.pipe(first(), mapTo(true)).toPromise();
+  public async pluginIsSetup(): Promise<boolean> {
+    // use deps and config as a cached resolver
+    if (this.pluginSetupDeps && this.config) {
+      return true;
+    }
+    return await this.pluginSetup$.pipe(take(2)).toPromise(); // once for pluginSetupDeps (sync) and twice for config (async)
+  }
+
+  public async pluginHasStarted(): Promise<boolean> {
+    return await this.getPluginStartDeps().then(() => true);
   }
 
-  public setBrowserDriverFactory(browserDriverFactory: HeadlessChromiumDriverFactory) {
-    this.browserDriverFactory = browserDriverFactory;
+  public setConfig(config: ReportingConfig) {
+    this.config = config;
+    this.pluginSetup$.next(true);
```

Comment:
Full "plugin setup" now depends on 2 things: the `pluginSetup` deps, and the async config. That is why this.pluginSetup$.next is called twice in this file.